File: evernote_client/enml/to_enml.py
"""Markdown to ENML conversion."""

from __future__ import annotations

import hashlib
import logging
import mimetypes
import re
from collections.abc import Callable
from pathlib import Path

from evernote_client.enml.types import Attachment, EnmlResult, ResourceInfo

logger = logging.getLogger(__name__)

ENML_HEADER = (
    '<?xml version="1.0" encoding="UTF-8"?>'
    '<!DOCTYPE en-note SYSTEM "http://xml.evernote.com/pub/enml2.dtd">'
    "<en-note>"
)
ENML_FOOTER = "</en-note>"

# --- Compiled regex patterns ---

_RE_HEADING = re.compile(r"^(#{1,6})\s+(.+)$")
_RE_HR = re.compile(r"^---+\s*$")
_RE_CHECKBOX = re.compile(r"^-\s+\[([ xX])\]\s+(.+)$")
_RE_UL_ITEM = re.compile(r"^[-*]\s+(.+)$")
_RE_UL_PREFIX = re.compile(r"^[-*]\s+")
_RE_OL_ITEM = re.compile(r"^\d+\.\s+(.+)$")
_RE_OL_PREFIX = re.compile(r"^\d+\.\s+")
_RE_CODE_FENCE_OPEN = re.compile(r"^```")
_RE_CODE_FENCE_CLOSE = re.compile(r"^```\s*$")
_RE_TABLE_ROW = re.compile(r"^\|.+\|")
_RE_TABLE_SEPARATOR = re.compile(r"^:?-+:?$")
_RE_BOLD = re.compile(r"\*\*(.+?)\*\*")
_RE_ITALIC = re.compile(r"\*(.+?)\*")
_RE_LINK = re.compile(r"\[([^\]]+)\]\(([^)]+)\)")
_RE_INLINE_CODE = re.compile(r"`([^`]+)`")
_RE_IMAGE = re.compile(r"!\[([^\]]*)\]\(([^)]+)\)")
_RE_EVERNOTE_RESOURCE = re.compile(r"^evernote-resource:([0-9a-f]{32})$")


def markdown_to_enml(
    md: str,
    existing_resources: list[ResourceInfo] | None = None,
) -> EnmlResult:
    """Convert Markdown to valid ENML for note creation."""
    if not md:
        return EnmlResult(enml=f"{ENML_HEADER}{ENML_FOOTER}")

    existing_map: dict[str, ResourceInfo] = (
        {r.hash_hex: r for r in existing_resources} if existing_resources else {}
    )
    attachments: list[Attachment] = []
    seen_hashes: set[str] = set()

    lines = md.split("\n")
    enml_parts: list[str] = [ENML_HEADER]
    i = 0

    while i < len(lines):
        line = lines[i]

        for parser in _BLOCK_PARSERS:
            result, consumed = parser(lines, i, existing_map, attachments, seen_hashes)
            if result is not None:
                enml_parts.append(result)
                i += consumed
                break
        else:
            if line.strip():
                # Check if line is a standalone image
                img_result = _parse_image_line(
                    line, existing_map, attachments, seen_hashes
                )
                if img_result is not None:
                    enml_parts.append(img_result)
                else:
                    text = _escape_xml(line)
                    text = _inline_md_to_enml(
                        text, existing_map, attachments, seen_hashes
                    )
                    enml_parts.append(f"<div>{text}</div>")
            i += 1

    enml_parts.append(ENML_FOOTER)
    return EnmlResult(enml="".join(enml_parts), attachments=attachments)


# --- Image rendering ---


def _render_image(
    alt: str,
    url: str,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> str:
    """Render a Markdown image to ENML."""
    # evernote-resource:<hash> scheme
    m = _RE_EVERNOTE_RESOURCE.match(url)
    if m:
        hash_hex = m.group(1)
        info = existing_map.get(hash_hex)
        if info:
            return f'<en-media type="{info.mime_type}" hash="{hash_hex}"/>'
        # Unknown resource — fall back to link
        display = alt or hash_hex
        return f'<a href="evernote-resource:{hash_hex}">{_escape_xml(display)}</a>'

    # HTTP/HTTPS — render as link, not downloadable attachment
    if url.startswith("http://") or url.startswith("https://"):
        display = _escape_xml(alt or url)
        return f'<a href="{_escape_xml(url)}">{display}</a>'

    # Local file
    path = _resolve_local_path(url)
    if path is None or not path.exists():
        logger.warning("Image path not found: %s", url)
        display = _escape_xml(alt or url)
        return f'<a href="{_escape_xml(url)}">{display}</a>'

    try:
        data = path.read_bytes()
    except OSError as exc:
        logger.warning("Cannot read image file %s: %s", path, exc)
        display = _escape_xml(alt or url)
        return f'<a href="{_escape_xml(url)}">{display}</a>'

    hash_bytes = hashlib.md5(data).digest()
    hash_hex = hash_bytes.hex()
    mime_type, _ = mimetypes.guess_type(path.name)
    mime_type = mime_type or "application/octet-stream"

    if hash_hex not in seen_hashes:
        seen_hashes.add(hash_hex)
        attachments.append(
            Attachment(
                hash_hex=hash_hex,
                hash_bytes=hash_bytes,
                mime_type=mime_type,
                data=data,
                filename=path.name,
                source_path=str(path),
            )
        )

    return f'<en-media type="{mime_type}" hash="{hash_hex}"/>'


def _resolve_local_path(url: str) -> Path | None:
    """Resolve a URL or path string to a Path, or return None."""
    if url.startswith("file://"):
        url = url[7:]
    try:
        return Path(url).expanduser()
    except Exception:
        return None


def _parse_image_line(
    line: str,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> str | None:
    """If the line is *only* a Markdown image, return ENML for it."""
    stripped = line.strip()
    m = _RE_IMAGE.fullmatch(stripped)
    if not m:
        return None
    alt, url = m.group(1), m.group(2)
    return _render_image(alt, url, existing_map, attachments, seen_hashes)


# --- Block parser infrastructure ---

_BlockParser = Callable[
    [list[str], int, dict[str, ResourceInfo], list[Attachment], set[str]],
    tuple[str | None, int],
]


def _single(
    fn: Callable[[str], str | None],
) -> _BlockParser:
    """Adapt a single-line parser (no resource state) to the full signature."""

    def wrapper(
        lines: list[str],
        i: int,
        existing_map: dict[str, ResourceInfo],
        attachments: list[Attachment],
        seen_hashes: set[str],
    ) -> tuple[str | None, int]:
        result = fn(lines[i])
        return (result, 1) if result is not None else (None, 0)

    return wrapper


def _single_inline(
    fn: Callable[
        [str, dict[str, ResourceInfo], list[Attachment], set[str]], str | None
    ],
) -> _BlockParser:
    """Adapt a single-line parser that receives resource state."""

    def wrapper(
        lines: list[str],
        i: int,
        existing_map: dict[str, ResourceInfo],
        attachments: list[Attachment],
        seen_hashes: set[str],
    ) -> tuple[str | None, int]:
        result = fn(lines[i], existing_map, attachments, seen_hashes)
        return (result, 1) if result is not None else (None, 0)

    return wrapper


# --- Block parsers ---


def _parse_heading(
    line: str,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> str | None:
    heading_match = _RE_HEADING.match(line)
    if not heading_match:
        return None
    level = len(heading_match.group(1))
    text = _escape_xml(heading_match.group(2))
    text = _inline_md_to_enml(text, existing_map, attachments, seen_hashes)
    return f"<h{level}>{text}</h{level}>"


def _parse_hr(line: str) -> str | None:
    if _RE_HR.match(line):
        return "<hr/>"
    return None


def _parse_checkbox(
    line: str,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> str | None:
    checkbox_match = _RE_CHECKBOX.match(line)
    if not checkbox_match:
        return None
    checked = checkbox_match.group(1).lower() == "x"
    text = _escape_xml(checkbox_match.group(2))
    text = _inline_md_to_enml(text, existing_map, attachments, seen_hashes)
    if checked:
        return f'<div><en-todo checked="true"/>{text}</div>'
    return f"<div><en-todo/>{text}</div>"


def _parse_list(
    tag: str,
    item_re: re.Pattern[str],
    prefix_re: re.Pattern[str],
    lines: list[str],
    i: int,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> tuple[str | None, int]:
    if not item_re.match(lines[i]):
        return None, 0
    items: list[str] = []
    start = i
    while i < len(lines) and prefix_re.match(lines[i]):
        m = item_re.match(lines[i])
        if m:
            items.append(_escape_xml(m.group(1)))
        i += 1
    parts = [f"<{tag}>"]
    for item in items:
        inner = _inline_md_to_enml(item, existing_map, attachments, seen_hashes)
        parts.append(f"<li>{inner}</li>")
    parts.append(f"</{tag}>")
    return "".join(parts), i - start


def _parse_ul(
    lines: list[str],
    i: int,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> tuple[str | None, int]:
    return _parse_list(
        "ul",
        _RE_UL_ITEM,
        _RE_UL_PREFIX,
        lines,
        i,
        existing_map,
        attachments,
        seen_hashes,
    )


def _parse_ol(
    lines: list[str],
    i: int,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> tuple[str | None, int]:
    return _parse_list(
        "ol",
        _RE_OL_ITEM,
        _RE_OL_PREFIX,
        lines,
        i,
        existing_map,
        attachments,
        seen_hashes,
    )


def _parse_code_block(
    lines: list[str],
    i: int,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> tuple[str | None, int]:
    if not _RE_CODE_FENCE_OPEN.match(lines[i]):
        return None, 0
    start = i
    i += 1
    content_lines: list[str] = []
    while i < len(lines):
        if _RE_CODE_FENCE_CLOSE.match(lines[i]):
            i += 1
            break
        content_lines.append(lines[i])
        i += 1
    content = _escape_xml("\n".join(content_lines))
    return f"<pre><code>{content}</code></pre>", i - start


def _parse_table(
    lines: list[str],
    i: int,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> tuple[str | None, int]:
    if not _RE_TABLE_ROW.match(lines[i]):
        return None, 0
    start = i
    rows: list[list[str]] = []
    while i < len(lines) and _RE_TABLE_ROW.match(lines[i]):
        line = lines[i].strip()
        cells = [c.strip() for c in line.strip("|").split("|")]
        # Skip separator row
        if all(_RE_TABLE_SEPARATOR.match(c) for c in cells):
            i += 1
            continue
        rows.append(cells)
        i += 1
    if not rows:
        return None, 0
    parts = ["<table>"]
    for idx, row in enumerate(rows):
        tag = "th" if idx == 0 else "td"
        cell_parts: list[str] = []
        for c in row:
            inner = _inline_md_to_enml(
                _escape_xml(c), existing_map, attachments, seen_hashes
            )
            cell_parts.append(f"<{tag}>{inner}</{tag}>")
        cells_html = "".join(cell_parts)
        parts.append(f"<tr>{cells_html}</tr>")
    parts.append("</table>")
    return "".join(parts), i - start


_BLOCK_PARSERS: list[_BlockParser] = [
    _parse_code_block,
    _parse_table,
    _single_inline(_parse_heading),
    _single(_parse_hr),
    _single_inline(_parse_checkbox),
    _parse_ul,
    _parse_ol,
]


# --- Helpers ---


def _escape_xml(text: str) -> str:
    """Escape XML special characters.

    Note: this does not handle already-escaped entities — passing ``&amp;``
    will produce ``&amp;amp;``.  Callers are expected to provide raw text.
    """
    text = text.replace("&", "&amp;")
    text = text.replace("<", "&lt;")
    text = text.replace(">", "&gt;")
    return text.replace('"', "&quot;")


def _inline_md_to_enml(
    text: str,
    existing_map: dict[str, ResourceInfo],
    attachments: list[Attachment],
    seen_hashes: set[str],
) -> str:
    """Convert inline markdown formatting to ENML tags.

    Operates on already XML-escaped text, so we match escaped markers.
    Images (![alt](url)) are processed first.
    """

    # Images: ![alt](url) — must come before link processing
    def replace_image(m: re.Match[str]) -> str:
        alt = m.group(1)
        url = m.group(2)
        return _render_image(alt, url, existing_map, attachments, seen_hashes)

    text = _RE_IMAGE.sub(replace_image, text)
    # Bold: **text**
    text = _RE_BOLD.sub(r"<b>\1</b>", text)
    # Italic: *text*
    text = _RE_ITALIC.sub(r"<i>\1</i>", text)
    # Links: [text](url) — url is XML-escaped so ( and ) are literal
    text = _RE_LINK.sub(r'<a href="\2">\1</a>', text)
    # Inline code: `text`
    return _RE_INLINE_CODE.sub(r"<code>\1</code>", text)

File: src/enml/to-enml.ts
import { createHash } from 'node:crypto';
import { existsSync, readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { basename } from 'node:path';
import { logger } from '../logger.js';
import type { Attachment, EnmlResult, ResourceInfo } from './types.js';

const ENML_HEADER =
  '<?xml version="1.0" encoding="UTF-8"?>' +
  '<!DOCTYPE en-note SYSTEM "http://xml.evernote.com/pub/enml2.dtd">' +
  '<en-note>';
const ENML_FOOTER = '</en-note>';

// --- Regex patterns ---
const RE_HEADING = /^(#{1,6})\s+(.+)$/;
const RE_HR = /^---+\s*$/;
const RE_CHECKBOX = /^-\s+\[([ xX])\]\s+(.+)$/;
const RE_UL_ITEM = /^[-*]\s+(.+)$/;
const RE_UL_PREFIX = /^[-*]\s+/;
const RE_OL_ITEM = /^\d+\.\s+(.+)$/;
const RE_OL_PREFIX = /^\d+\.\s+/;
const RE_CODE_FENCE_OPEN = /^```/;
const RE_CODE_FENCE_CLOSE = /^```\s*$/;
const RE_TABLE_ROW = /^\|.+\|/;
const RE_TABLE_SEPARATOR = /^:?-+:?$/;
// Bold must be applied before italic so `**bold**` isn't consumed as italic.
// Limitation: nested patterns like `*italic **bold** more italic*` will
// produce incorrect nesting — a full markdown parser would be needed to fix.
const RE_BOLD = /\*\*(.+?)\*\*/g;
const RE_ITALIC = /\*(.+?)\*/g;
const RE_LINK = /\[([^\]]+)\]\(([^)]+)\)/g;
const RE_INLINE_CODE = /`([^`]+)`/g;
const RE_IMAGE = /!\[([^\]]*)\]\(([^)]+)\)/g;
const RE_IMAGE_FULL = /^!\[([^\]]*)\]\(([^)]+)\)$/;
const RE_EVERNOTE_RESOURCE = /^evernote-resource:([0-9a-f]{32})$/;

export function markdownToEnml(
  md: string,
  existingResources?: ResourceInfo[]
): EnmlResult {
  if (!md) return { enml: `${ENML_HEADER}${ENML_FOOTER}`, attachments: [] };

  const existingMap = new Map<string, ResourceInfo>();
  if (existingResources) {
    for (const r of existingResources) existingMap.set(r.hashHex, r);
  }

  const attachments: Attachment[] = [];
  const seenHashes = new Set<string>();
  const lines = md.split('\n');
  const enmlParts: string[] = [ENML_HEADER];
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];

    // Try block parsers in order
    let consumed = 0;
    let result: string | null = null;

    // Code block
    ({ result, consumed } = parseCodeBlock(lines, i));
    if (result !== null) {
      enmlParts.push(result);
      i += consumed;
      continue;
    }

    // Table
    ({ result, consumed } = parseTable(
      lines,
      i,
      existingMap,
      attachments,
      seenHashes
    ));
    if (result !== null) {
      enmlParts.push(result);
      i += consumed;
      continue;
    }

    // Heading
    result = parseHeading(line, existingMap, attachments, seenHashes);
    if (result !== null) {
      enmlParts.push(result);
      i++;
      continue;
    }

    // HR
    if (RE_HR.test(line)) {
      enmlParts.push('<hr/>');
      i++;
      continue;
    }

    // Checkbox
    result = parseCheckbox(line, existingMap, attachments, seenHashes);
    if (result !== null) {
      enmlParts.push(result);
      i++;
      continue;
    }

    // Unordered list
    ({ result, consumed } = parseList(
      'ul',
      RE_UL_ITEM,
      RE_UL_PREFIX,
      lines,
      i,
      existingMap,
      attachments,
      seenHashes
    ));
    if (result !== null) {
      enmlParts.push(result);
      i += consumed;
      continue;
    }

    // Ordered list
    ({ result, consumed } = parseList(
      'ol',
      RE_OL_ITEM,
      RE_OL_PREFIX,
      lines,
      i,
      existingMap,
      attachments,
      seenHashes
    ));
    if (result !== null) {
      enmlParts.push(result);
      i += consumed;
      continue;
    }

    // Regular line
    if (line.trim()) {
      // Check for standalone image
      const imgResult = parseImageLine(
        line,
        existingMap,
        attachments,
        seenHashes
      );
      if (imgResult !== null) {
        enmlParts.push(imgResult);
      } else {
        let text = escapeXml(line);
        text = inlineMdToEnml(text, existingMap, attachments, seenHashes);
        enmlParts.push(`<div>${text}</div>`);
      }
    }
    i++;
  }

  enmlParts.push(ENML_FOOTER);
  return { enml: enmlParts.join(''), attachments };
}

// --- Image rendering ---

function renderImage(
  alt: string,
  url: string,
  existingMap: Map<string, ResourceInfo>,
  attachments: Attachment[],
  seenHashes: Set<string>
): string {
  // evernote-resource:<hash> scheme
  const m = RE_EVERNOTE_RESOURCE.exec(url);
  if (m) {
    const hashHex = m[1];
    const info = existingMap.get(hashHex);
    if (info) {
      if (info.data && !seenHashes.has(hashHex)) {
        seenHashes.add(hashHex);
        attachments.push({
          hashHex,
          hashBytes: new Uint8Array(Buffer.from(hashHex, 'hex')),
          mimeType: info.mimeType,
          data: info.data,
          filename: info.filename,
          sourcePath: '',
        });
      }
      return `<en-media type="${info.mimeType}" hash="${hashHex}"/>`;
    }
    // Resource not available — text placeholder
    const display = alt || hashHex;
    return `[image: ${escapeXml(display)}]`;
  }

  // HTTP/HTTPS — render as link
  if (url.startsWith('http://') || url.startsWith('https://')) {
    const display = escapeXml(alt || url);
    return `<a href="${escapeXml(url)}">${display}</a>`;
  }

  // Local file
  const path = resolveLocalPath(url);
  if (!path || !existsSync(path)) {
    logger.warn(`Image path not found: ${url}`);
    const display = escapeXml(alt || url);
    return `<a href="${escapeXml(url)}">${display}</a>`;
  }

  let data: Uint8Array;
  try {
    data = readFileSync(path);
  } catch (err) {
    logger.warn(`Cannot read image file ${path}: ${err}`);
    const display = escapeXml(alt || url);
    return `<a href="${escapeXml(url)}">${display}</a>`;
  }

  const hashBytes = createHash('md5').update(data).digest();
  const hashHex = hashBytes.toString('hex');

  // Guess mime type
  const ext = path.split('.').pop()?.toLowerCase() ?? '';
  const mimeMap: Record<string, string> = {
    png: 'image/png',
    jpg: 'image/jpeg',
    jpeg: 'image/jpeg',
    gif: 'image/gif',
    webp: 'image/webp',
    svg: 'image/svg+xml',
    pdf: 'application/pdf',
  };
  const mimeType = mimeMap[ext] ?? 'application/octet-stream';

  if (!seenHashes.has(hashHex)) {
    seenHashes.add(hashHex);
    attachments.push({
      hashHex,
      hashBytes: new Uint8Array(hashBytes),
      mimeType,
      data: new Uint8Array(data),
      filename: basename(path),
      sourcePath: path,
    });
  }

  return `<en-media type="${mimeType}" hash="${hashHex}"/>`;
}

function resolveLocalPath(url: string): string | null {
  if (url.startsWith('file://')) url = url.substring(7);
  try {
    return resolve(
      url.startsWith('~') ? url.replace('~', process.env.HOME ?? '') : url
    );
  } catch {
    return null;
  }
}

function parseImageLine(
  line: string,
  existingMap: Map<string, ResourceInfo>,
  attachments: Attachment[],
  seenHashes: Set<string>
): string | null {
  const m = RE_IMAGE_FULL.exec(line.trim());
  if (!m) return null;
  const rendered = renderImage(
    m[1],
    m[2],
    existingMap,
    attachments,
    seenHashes
  );
  return `<div>${rendered}</div>`;
}

// --- Block parsers ---

function parseHeading(
  line: string,
  existingMap: Map<string, ResourceInfo>,
  attachments: Attachment[],
  seenHashes: Set<string>
): string | null {
  const m = RE_HEADING.exec(line);
  if (!m) return null;
  const level = m[1].length;
  let text = escapeXml(m[2]);
  text = inlineMdToEnml(text, existingMap, attachments, seenHashes);
  return `<h${level}>${text}</h${level}>`;
}

function parseCheckbox(
  line: string,
  existingMap: Map<string, ResourceInfo>,
  attachments: Attachment[],
  seenHashes: Set<string>
): string | null {
  const m = RE_CHECKBOX.exec(line);
  if (!m) return null;
  const checked = m[1].toLowerCase() === 'x';
  let text = escapeXml(m[2]);
  text = inlineMdToEnml(text, existingMap, attachments, seenHashes);
  return checked
    ? `<div><en-todo checked="true"/>${text}</div>`
    : `<div><en-todo/>${text}</div>`;
}

function parseList(
  tag: string,
  itemRe: RegExp,
  prefixRe: RegExp,
  lines: string[],
  i: number,
  existingMap: Map<string, ResourceInfo>,
  attachments: Attachment[],
  seenHashes: Set<string>
): { result: string | null; consumed: number } {
  if (!itemRe.test(lines[i])) return { result: null, consumed: 0 };
  const items: string[] = [];
  const start = i;
  while (i < lines.length && prefixRe.test(lines[i])) {
    const m = itemRe.exec(lines[i]);
    if (m) items.push(escapeXml(m[1]));
    i++;
  }
  const parts = [`<${tag}>`];
  for (const item of items) {
    const inner = inlineMdToEnml(item, existingMap, attachments, seenHashes);
    parts.push(`<li>${inner}</li>`);
  }
  parts.push(`</${tag}>`);
  return { result: parts.join(''), consumed: i - start };
}

function parseCodeBlock(
  lines: string[],
  i: number
): { result: string | null; consumed: number } {
  if (!RE_CODE_FENCE_OPEN.test(lines[i])) return { result: null, consumed: 0 };
  const start = i;
  i++;
  const contentLines: string[] = [];
  while (i < lines.length) {
    if (RE_CODE_FENCE_CLOSE.test(lines[i])) {
      i++;
      break;
    }
    contentLines.push(lines[i]);
    i++;
  }
  const content = escapeXml(contentLines.join('\n'));
  return {
    result: `<pre><code>${content}</code></pre>`,
    consumed: i - start,
  };
}

function parseTable(
  lines: string[],
  i: number,
  existingMap: Map<string, ResourceInfo>,
  attachments: Attachment[],
  seenHashes: Set<string>
): { result: string | null; consumed: number } {
  if (!RE_TABLE_ROW.test(lines[i])) return { result: null, consumed: 0 };
  const start = i;
  const rows: string[][] = [];
  while (i < lines.length && RE_TABLE_ROW.test(lines[i])) {
    const line = lines[i].trim();
    const cells = line
      .replace(/^\||\|$/g, '')
      .split('|')
      .map((c) => c.trim());
    if (cells.every((c) => RE_TABLE_SEPARATOR.test(c))) {
      i++;
      continue;
    }
    rows.push(cells);
    i++;
  }
  if (rows.length === 0) return { result: null, consumed: 0 };

  const parts = ['<table>'];
  for (let idx = 0; idx < rows.length; idx++) {
    const tag = idx === 0 ? 'th' : 'td';
    const cellParts: string[] = [];
    for (const c of rows[idx]) {
      const inner = inlineMdToEnml(
        escapeXml(c),
        existingMap,
        attachments,
        seenHashes
      );
      cellParts.push(`<${tag}>${inner}</${tag}>`);
    }
    parts.push(`<tr>${cellParts.join('')}</tr>`);
  }
  parts.push('</table>');
  return { result: parts.join(''), consumed: i - start };
}

// --- Helpers ---

function escapeXml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}

function inlineMdToEnml(
  text: string,
  existingMap: Map<string, ResourceInfo>,
  attachments: Attachment[],
  seenHashes: Set<string>
): string {
  // Images first
  text = text.replace(RE_IMAGE, (_match, alt: string, url: string) =>
    renderImage(alt, url, existingMap, attachments, seenHashes)
  );
  // Bold
  text = text.replace(RE_BOLD, '<b>$1</b>');
  // Italic
  text = text.replace(RE_ITALIC, '<i>$1</i>');
  // Links
  text = text.replace(RE_LINK, '<a href="$2">$1</a>');
  // Inline code
  text = text.replace(RE_INLINE_CODE, '<code>$1</code>');
  return text;
}
